lwotg/flows.go

```diff
@@ -0,0 +1,112 @@
+// When called it should begin tranmitting packets using the tx FlowController channels, and
+// receiving them using the rx FlowController. Tx and Rx should be spawned into new goroutines
+// and the function should return. Telemetry that is published by the flow should be written
+// to the corresponding FlowController GNMI channel. Encountered errors should be written to
+// the error channel.
+type TXRXFn func(tx, rx *FlowController)
+
+// FlowController acts as the control mechanism for a specific direction of a specific flow.
+type FlowController struct {
+	// Stop is a channel used to indicate that the function that has been called should
+	// cease to send or receive packets.
+	Stop chan struct{}
+}
+
+// NewFlowController returns an initialised FlowController.
+func NewFlowController() *FlowController {
+	return &FlowController{
+		Stop: make(chan struct{}),
+	}
+}
+
+// FlowGeneratorFn is a function that takes an input OTG Flow and, if it is able to, returns
+// a TXRXFn that controls that flow. If the function is not able to handle the flow (e.g., it
+// is an IP flow and the function can handle only MPLS) it returns no TXRXFn, and sets the
+// bool return argument to false. If the function can handle the flow, it returns a TXRXFn that
+// can be used to start the flow, with the bool return argument set to true. Errors encountered
+// whilst generating a TXRXFn for a flow that the function can return should be returned and
+// are treated as errors that should be returned to the caller.
+//
+// The arguments to a FlowGeneratorFn are as follows:
+//   - otg.Flow - the flow that is to be generated as described by the OTG schema.
+//   - []*OTGIntf - information as to the set of interfaces that are currently within the OTG
+//     configuration. This information can be used to map the flow creation functionality to
+//     the underlying interface.
+type FlowGeneratorFn func(*otg.Flow, []*OTGIntf) (TXRXFn, bool, error)
+
+// AddFlowHandlers adds the set of flow generator functions specified to the flow handlers that
+// are considered as candidates by the server.
+func (s *Server) AddFlowHandlers(fns ...FlowGeneratorFn) {
+	s.fhMu.Lock()
+	defer s.fhMu.Unlock()
+	s.flowHandlers = append(s.flowHandlers, fns...)
+}
+
+// handleFlows takes the set of flows provided and returns the TXRX functions that control them.
+// It returns an error if a flow cannot be handled by the registered set of flow handlers.
+func (s *Server) handleFlows(flows []*otg.Flow) ([]TXRXFn, error) {
+	s.fhMu.Lock()
+	defer s.fhMu.Unlock()
+
+	flowMethods := []TXRXFn{}
+	intfs := s.interfaces()
+	for _, flow := range flows {
+		var handled bool
+		for _, fn := range s.flowHandlers {
+			txrx, ok, err := fn(flow, intfs)
+			switch {
+			case err != nil:
+				// The flow could be handled, but an error occurred.
+				return nil, status.Errorf(codes.Internal, "error generating flows, %v", err)
```

Comment:
Would it be worth trying to see if another flowHandler can handle the flow before returning the error?

Reply:
That's a good question -- I was assuming here that the pattern is that there is a slice of flowHandlers that are each considered as to whether they can map the flow. If a handler says "yes, I can handle this, but there is an error in this IPv4 address" (for example), then this would be an error that is returned, whereas the case of "I don't handle flows that have randomised IPv4 source addresses" would not be an error, but a return of nil with the "ok" return arg set to `false`.

Is this intuitive to you? If it doesn't make sense then it might be better just to say that flows run through all handlers, and if none returns (for erroneous or otherwise reasons) then we return an error. We'd probably want to log that there was a problem in some handler though that /should/ have been able to handle the flow, but couldn't.